test/minishift/test_suite.go

```diff
@@ -0,0 +1,161 @@
+	"github.com/stretchr/testify/assert"
+	"net/http"
+	"os"
+	"sync"
+	"testing"
+	"time"
+)
+
+// TestSuite is a base for tests using Minishift and gorm db
+type TestSuite struct {
+	gormsupport.DBTestSuite
+	ClusterService      *stub.ClusterService
+	AuthService         *stub.AuthService
+	Config              *configuration.Data
+	toReset             func()
+	minishiftUrl        string
+	minishiftAdminName  string
+	minishiftAdminToken string
+	minishiftUserName   string
+	minishiftUserToken  string
+}
+
+func (s *TestSuite) SetupTest() {
+	resource.Require(s.T(), resource.Database, "F8_MINISHIFT")
+	s.minishiftUrl = os.Getenv("F8_MINISHIFT_URL")
+	s.minishiftAdminName = os.Getenv("F8_MINISHIFT_ADMIN_NAME")
+	s.minishiftAdminToken = os.Getenv("F8_MINISHIFT_ADMIN_TOKEN")
+	s.minishiftUserName = os.Getenv("F8_MINISHIFT_USER_NAME")
+	s.minishiftUserToken = os.Getenv("F8_MINISHIFT_USER_TOKEN")
+	s.DBTestSuite.SetupTest()
+	s.Config, s.toReset = prepareConfig(s.T())
+
+	log.InitializeLogger(s.Config.IsLogJSON(), s.Config.GetLogLevel())
+
+	s.ClusterService = &stub.ClusterService{
+		APIURL: s.minishiftUrl,
+		User:   s.minishiftAdminName,
+		Token:  s.minishiftAdminToken,
+	}
+	s.AuthService = &stub.AuthService{
+		OpenShiftUsername:  s.minishiftUserName,
+		OpenShiftUserToken: s.minishiftUserToken,
+	}
+}
+
+func (s *TestSuite) TearDownTest() {
+	s.DBTestSuite.TearDownTest()
+	s.toReset()
+}
+
+func (s *TestSuite) GetClusterService() cluster.Service {
+	return s.ClusterService
+}
+
+func (s *TestSuite) GetAuthService(tenantID uuid.UUID) auth.Service {
+	s.AuthService.TenantID = tenantID
+	return s.AuthService
+}
+
+func (s *TestSuite) GetConfig() *configuration.Data {
+	return s.Config
+}
+
+func prepareConfig(t *testing.T) (*configuration.Data, func()) {
+	resetVars := test.SetEnvironments(
+		test.Env("F8_AUTH_TOKEN_KEY", "foo"),
+		test.Env("F8_API_SERVER_USE_TLS", "false"),
+		test.Env("F8_LOG_LEVEL", "error"),
+		test.Env("F8_KEYCLOAK_URL", "http://keycloak.url.com"))
+	config, resetConf := test.LoadTestConfig(t)
+	reset := func() {
+		resetVars()
+		resetConf()
+	}
+	return config, reset
+}
+
+func VerifyObjectsPresence(t *testing.T, mappedObjects map[string]environment.Objects, options openshift.ApplyOptions, version string) {
+	for ns, objects := range mappedObjects {
+		var wg sync.WaitGroup
+		for _, obj := range objects {
+			wg.Add(1)
+			go func(toMarkAsDone *sync.WaitGroup, t *testing.T) {
+				iterations := 0
+				defer toMarkAsDone.Done()
+				for {
+					if openshift.IsOfKind(environment.ValKindProjectRequest, environment.ValKindProject, environment.ValKindNamespace)(obj) {
+						if environment.GetKind(obj) == environment.ValKindProjectRequest {
+							obj["kind"] = environment.ValKindNamespace
+						}
+						response, err := openshift.Apply(obj, "GET", options)
+						if err == nil && environment.HasValidStatus(response) {
+							break
+						} else if iterations >= 20 {
+							assert.NoError(t, err)
```

Comment:
If I understand correctly, you're waiting for up to 20 iterations and if there's still an error, then the test will fail ? As suggested above, I believe that the code would be easier to read if you use a `select channels` with a timeout, instead.

Reply:
I'm not sure if `channels` is not an overkill:
* I don't need to store all errors - it is enough just the last one
* I don't need to distinguish between timeout and error
* the likelihood that the error would occur is low - at least I haven't received any, I check the error just for sure
* the most important check is `environment.HasValidStatus(response)`

Reply:
I'll try to improve it ;-)

Reply:
improved as part of 0093df3 
see here https://github.com/fabric8-services/fabric8-tenant/pull/685/commits/0093df3ccbb9f9bc8a1d000ea6aec5510bab03c4#diff-e526609efd4c9f26c792830751f918e4R103